### ISSUE TITLE
Multiple related persons combining emergency contacts and referents

### ISSUE THREAD

**Issue by lamontfr** [savoirfairelinux/sous-chef#740]
(As requested by and discussed with Jean-François on 2017-05-12, multiple referents are essential. This is a solution that minimizes changes such that we can finish milestone 1.1 ASAP)

(EDITED by @lamontfr 2017-05-20 19:12 EDT)
### Expected Behaviour
- For each instance of a "Related person" the attributes will be displayed as follows in the second block of Client / Information / Personal with block title = Related Persons :
  - First name / Prénom (required)
  - Last name  / Nom (required)
  - Nature of relationship / Lien de relation (required, free text) (ex. Doctor, Friend, Gov. official, Daughter)
  - Home Phone / Téléphone maison (optional)
  - Cell. Phone / Téléphone cellulaire (optional)
  - Emergency Contact / Contact d'urgence : [x] (checked if YES)
  - Referent / Référent : [x] (checked if YES)
  - Workplace information / Lieu de travail (required if referent checked, otherwise empty)
  - Referral Date / Date du référencement  (required if referent checked, otherwise empty)
  - Referral reason / Raison du référencement  (required if referent checked, otherwise empty)
  - Remark / Remarque (optional, used by staff)
- We may have none or many "Related persons" for a client.
- A member may simultaneously be a referent and an emergency contact.
- A member may be a related person but neither a referent nor an emergency contact.
- There is no requirement to supply an EmergencyContact (or a referent) in the new
 client wizard.

### Actual Behaviour
- Can only have one referent per client.
- Emergency contact is compulsory

### Revised suggestion to solve issue (based on comments)
- Create a new model Relationship and remove models Referencing and EmergencyContact and move data from the two latter to the former.
- Relationship would be : 
```python
class Relationship(models.Model):
    member = models.ForeignKey('member.Member')
    client = models.ForeignKey('member.Client')
    nature = models.CharField(max_length=100)
    type = JSONField()
    extra_fields = JSONField()
    remark = models.TextField()
```
- field "type" is a multiple choice : none, one or many of ('referent', 'emergency').
- when 'referent' is in field "type", extra_fields are : referral_date, referral_reason.
- In the UI use a new label "Related Persons" instead of "Emergency Contacts"
- Use the Django migration module to : 
  - Copy Referencing data into Relationship, including extra Referents migrated from Feast that does not show in the UI.
  - Copy EmergencyContact data into Relationship.
- In the UI remove Referent
- We dot not plan to run the data migration from Feast to Souschef again. Jean-François told me that they try to continously update the client data in Souschef (but not the orders). When we put milestone 1.1 in production they will run in parallel with Feast and manually adjust the discrepancies.

**Comment by lingxiaoyang** [comment]
@lamontfr  In fact, I think we have already supported this on data model level!

Right now, we have `member.Referencing` that has one FK to `member.Client` and another FK to `member.Member`. Thus it's already a many-to-many-style model! We can have multiple Referencing instances with the same Client and different Member. And it's only necessary to fix the form validation and display logic.

**Comment by lamontfr** [comment]
@lingxiaoyang I agree with your comment. However Jean-François showed me actual examples (in the current Feast system) of related persons that are neither referents nor emergency contacts. I forgot to mention this in the expected behavior. This was in the minutes or our meeting on 2017-05-10 and 2017-05-12, in section 1.6. I just forwarded them to you and will cc you in the future.

Here is a conceptual way to see the inheritance tree : 
- A RelatedPerson is a concrete (not abstract) superclass that has attributes : member.ClientFK, member.MemberFK, first_name,  last_name, relationship, phone, remark.
- A Referent is a concrete subclass of RelatedPerson that has extra attributes : referent_flag, date, organization, referral_reason.
- An EmergencyContact is a concrete subclass of RelatedPerson that has extra attribute : emergency_contact_flag
 
This is why I proposed the above solution, which amounts to have only one many-to-many model that can serve all purposes following this pattern :  https://martinfowler.com/eaaCatalog/singleTableInheritance.html. SInce a given RelatedPerson may be simultaneously a Referent and an EmergencyContact, this pattern can also support this if we have a single ORM model (not  subclassed models).

Your ideas about a better solution are welcome ! All I want is to minimize the work needed such that we can put Souschef in production at Roulant ASAP.

**Comment by lingxiaoyang** [comment]
@lamontfr  Thank you for the information. Yes, to merge referents and emergency contacts, it is necessary to delete `Referencing` and `EmergencyContact` models and move all data into the new relation model. 

In Django we may define it as:

```python
class Relationship(models.Model):
    member = models.ForeignKey('member.Member')
    client = models.ForeignKey('member.Client')
    type = models.CharField(choices=['referent', 'emergency'])
    extra_fields = JSONField()
```

1. Regarding the model name, I would not remove "Person" as it doesn't define a member or client, but how they are related.
2. Referent and emergency contact have very different fields. Thus I think it's better to put the extra fields as a JSON field and validate them using form validation. It's also easier if we extend more relationship types later.
3. First name, last name, phone and organization are already there on the related Member instance. 

Do we also need to update the `datamigration` module that moves data from Feast to Sous-Chef?

**Comment by lamontfr** [comment]
@lingxiaoyang I agree with your proposal but I have some comments.
1. We need more type choices to handle the cases I mentioned in my comment above (May 16 21:15 EDT). The model would then be : 
```python
class Relationship(models.Model):
    member = models.ForeignKey('member.Member')
    client = models.ForeignKey('member.Client')
    type = models.CharField(choices=['referent', 'emergency', 'referent+emergency', 'related'])
    extra_fields = JSONField()
```
2. Here 'related' means that the member knows the client, but is neither a referent nor an emergency contact. For example, the member may be a relative living far away.
3. We dot not plan to run the data migration from Feast to Souschef again. Jean-François told me that they try to update the client data in Souschef (but not the orders). When we put milestone 1.1 in production they will run in parallel with Feast and manually adjust the discrepancies.

Please let me know what you think about this.

**Comment by lamontfr** [comment]
@lingxiaoyang I agree that a JSONField with setters and getters would be more elegant for the multiple choices type field.

**Comment by lamontfr** [comment]
@lingxiaoyang I updated the Expected behavior and Suggestion for solution. Here is the REMOVED suggestion. : 
- Reuse model EmergencyContact by adding missing attributes
  - Boolean : Emergency Contact / Contact d'urgence : [x] (checked if YES)
  - Boolean : Referent / Référent : [x] (checked if YES)
  - Text : Organization / Organisation (required if referent checked, otherwise optional)
  - Date / Date   (required if referent checked, otherwise optional)
  - Text : Referral reason / Raison de la référence  (required if referent checked, otherwise empty)
  - Text : Remark / Remarque (optional, used by staff)
- In the UI use a new label "Related Persons"
- Write a script to copy Referent data into
  EmergencyContact, including extra Referents migrated from Feast that does not show in the UI.
- In the production database at SRoulant, the duplicate referents (instances of model Member having identical firstname+lastname and which are pointed to by instances of model Referencing) should be removed. For a given referent professional, a single Member instance should remain to which instances of EmergencyContact having Referent=Yes will point to. (This was moved from issue #698).
- Remove model Referencing and in the UI remove Referent
- Eventually rename model EmergencyContact to RelatedPerson

**Comment by lingxiaoyang** [comment]
@lamontfr  Yes. Thank you for the summary!

> Write a script to copy Referencing data into Relationship, including extra Referents migrated from Feast that does not show in the UI.
> Write a script to copy EmergencyContact data into Relationship.

This will be done in the Django migration module. An example is [here](https://github.com/savoirfairelinux/sous-chef/blob/dev/src/member/migrations/0026_change_to_emergency_contacts.py) and it's also possible to [unit test it](https://github.com/savoirfairelinux/sous-chef/blob/dev/src/member/tests.py#L3645).

**Comment by lamontfr** [comment]
@lingxiaoyang I edited the Expected behavior and removed : 
- In the production database at SRoulant, the duplicate referents (instances of model Member having identical firstname+lastname and which are pointed to by instances of model Referencing) should be removed. For a given referent professional, a single Member instance should remain to which instances of Relationship having type = ['Referent'] will point to. (This was moved from issue #698).

This has been moved to issue #749.
